docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,233 @@
+API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/).
+
+##### Security
+
+Worker Pods run a privileged container but are granted zero privilege on the API Server.
+
+##### kmod images
+
+It is worth noting that worker Pods pull the kmod image from registries **into their filesystem**, not the node's.
+That filesystem disappears as soon as the Pod is terminated.  
+This means that node image caching mechanisms cannot be leveraged, and any Pod creation will entail data transfers from
+the kmod image's registry to the Pod.
+
+This concern is mitigated by dropping the requirements on the kmod image.
+The worker image contains all the utilities required to load the kernel modules, and the kmod image is now merely a
+vehicle for `.ko` and `depmod` dependency files.  
+This results in much smaller images that can be based off `scratch` in most cases.
+
+##### Setting the firmware lookup path
+
+Worker Pods will be able to write the firmware lookup path to `/sys/module/firmware_class/parameters/path` each time
+they run.
+This behaviour will be controlled by an operator setting, which will be written as part of the worker configuration.
+
+#### High-level reconciliation loop
+
+```mermaid
+---
+High-level reconciliation loop
+---
+
+flowchart TD
+    RNMS[Reconcile NodeModuleState]
+    P[Worker Pod starts\nrunning on Node]
+    DMO{NodeModuleState\nspec matches status?}
+    TS{Last Node Ready\ntransition more recent than\nNodeModuleState status?}
+    E(((End)))
+
+    M{{Module update}} --> RNMS
+    N{{Node update}} --> RNMS
+
+    CR{{Controller starts}} --> TS
+    NR{{Node becomes Ready}} --> TS
+
+    RNMS --> DMO
+
+    DMO -- Yes --> E
+    DMO -- No --> P
+
+    TS -- Yes --> P
+    TS -- No --> E
+
+    P -- Pod completed or failed --> RO[Controller reads\nPod output] --> OUS[Controller updates\nNodeModuleState status] --> DMO
+```
+
+#### Device plugins
+
+KMM will still run the device plugin specified in the `Module` resource as a DaemonSet; however instead of relying on
+the `kmm.node.kubernetes.io/$namespace.$module.ready` label in its selector, we will introduce a new label
+that is independent of the kmod's state in the node.
+This should make it easier to terminate the device plugin before we try to unload the kmod.
+
+#### Image builds and kmod signing
+
+Those flows are not affected by this change.
+
+#### Addressing goals
+
+This design addresses the goals stated at the beginning of this proposal:
+
+1. **Improve the reliability of module loading and unloading**  
+   KMM now runs its own code in the pod, which gives us a lot more control in error handling.
+
+2. **Improve operator upgrade experience**  
+   Worker pods are one-off, short-lived objects.
+   We never try to update them and are not constrained by spec changes and immutable selectors.
+
+3. **Reduce the overall KMM footprint**  
+   We now run only one Pod per Node, for only a short amount of time.
```

Comment:
We do create a `NodeModuleState` resource for each module thought. Isn't it increasing the footprint more than what it could be with a single `NodeModulesState`?

Reply:
Proposal changed to one `NodeModulesState` per `Node`.